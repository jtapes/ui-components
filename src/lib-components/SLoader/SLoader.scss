@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

$loader-radius: 10 !default;

$length: $loader-radius * 2 * 3.14;

.s-loader {
  @include z-index(loader);

  position: absolute;
  display: flex;
  justify-content: center;
  width: 100%;
  color: $color--font-secondary;
  background-color: $color--background;
  inset: 0;

  &-spinner {
    transform: rotate(-90deg);

    &__path {
      stroke-width: 2px;
      animation: fill 1.5s ease-in-out infinite;
    }
  }

  &__wrapper {
    display: flex;
    align-items: center;
  }

  &--group {
    color: $color--primary-accent;
    background-color: $color--white;
    opacity: 0.84;

    .s-loader {
      &__wrapper {
        position: absolute;
        top: 5%;
      }

      &-spinner {
        width: $size--60;
        height: $size--60;
      }
    }
  }

  &--relative {
    position: relative;
    pointer-events: none;
  }

  &--input {
    background-color: $color--white;
  }

  @keyframes fill {
    0% {
      stroke-dasharray: 0, $length;
    }

    50% {
      stroke-dasharray: $length, 0;
      stroke-dashoffset: 0;
    }

    100% {
      stroke-dasharray: 0, $length;
      stroke-dashoffset: -$length;
    }
  }
}
