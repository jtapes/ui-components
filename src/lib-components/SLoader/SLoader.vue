<template>
  <transition name="s-fade" @after-leave="afterLeave">
    <div
      v-show="visible"
      class="s-loader"
      :class="{ 's-loader--group': group, 's-loader--input': input }"
    >
      <div v-show="visible" class="s-loader__wrapper">
        <svg
          ref="spinner"
          class="s-loader-spinner icon-32"
          viewBox="0 0 32 32"
        >
          <circle
            class="s-loader-spinner__path"
            cx="16"
            cy="16"
            r="10"
            fill="none"
            stroke="currentColor"
          />
        </svg>
        <div v-if="text" class="ml-24 bold">
          {{ text }}
        </div>
      </div>
    </div>
  </transition>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'

/** Компонент для директивы загрузки */
@Component
export default class SLoader extends Vue {
  visible = false
  group = false
  input = false
  text = ''

  afterLeave(): void {
    this.$emit('after-leave')
  }
}
</script>
