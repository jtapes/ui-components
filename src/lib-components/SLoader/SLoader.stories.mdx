import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Props,
  Preview,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SLoader from '@/lib-components/SLoader/SLoader.vue'
import Vue from 'vue'
import loader from '@/plugins/directives/loader.ts'

<Meta
  title="DESIGN SYSTEM/Atom/Loader"
  component={SLoader}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    loading: { control: 'boolean' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1128%3A88',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Loader

<Description of={SLoader} />

export const Template = (args, { argTypes }) => {
  Vue.use(loader)
  return {
    props: Object.keys(argTypes),
    components: {SLoader},
    template:
      `<div style="width: 100px; height: 100px;" v-loading="loading">
      Контент
    </div>`,
  }
}

export const TemplateGrouped = (args, { argTypes }) => {
  Vue.use(loader)
  return {
    props: Object.keys(argTypes),
    components: { SLoader },
    template:
      `<div style="width: 500px; height: 500px;" v-loading.group="{ loading, text: 'Формирование договора..' }">
        Контент
    </div>`,
  }
}

<Preview withToolbar withSource="none">
  <Story name="SLoader" args={{ loading: true }}>
    {Template.bind({})}
  </Story>
</Preview>


<Preview withToolbar withSource="none">
  <Story name="SLoader grouped" args={{ loading: true }}>
    {TemplateGrouped.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
    <div style="width: 100px; height: 100px;" v-loading="loading">
        Контент
    </div>
  `}
/>

<ArgsTable of={SLoader} />
