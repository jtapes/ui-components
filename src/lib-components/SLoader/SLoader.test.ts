import { mount } from '@vue/test-utils'
import SLoader from '@/lib-components/SLoader/SLoader.vue'

describe('SLoader', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SLoader)
    wrapper.vm.$data.visible = true
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Should change visible', async () => {
    const wrapper = mount(SLoader)

    const loader = wrapper.find('.s-loader')
    expect(loader.isVisible()).toBe(false)

    wrapper.vm.$data.visible = true
    await wrapper.vm.$nextTick()
    expect(loader.isVisible()).toBe(true)
  })
})
