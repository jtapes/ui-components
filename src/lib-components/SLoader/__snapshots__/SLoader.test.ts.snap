// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLoader Is Vue snapshot 1`] = `
<transition-stub
  name="s-fade"
>
  <div
    class="s-loader"
    style="display: none;"
  >
    <div
      class="s-loader__wrapper"
      style="display: none;"
    >
      <svg
        class="s-loader-spinner icon-32"
        viewBox="0 0 32 32"
      >
        <circle
          class="s-loader-spinner__path"
          cx="16"
          cy="16"
          fill="none"
          r="10"
          stroke="currentColor"
        />
      </svg>
       
      <!---->
    </div>
  </div>
</transition-stub>
`;
